File: Language Pad/Notebook/NotebookFile.vb
Imports System.IO
Imports TundraLib.ZiaFile
Imports System.IO.Compression

<Serializable()>
Public Class NotebookPage
    Public Title As String
    Public RTF As String
End Class

<Serializable()>
Public Class NotebookFile
    Private _Modified As Boolean = False

    Public Property Modified As Boolean
        Get
            Return _Modified
        End Get
        Set(value As Boolean)
            _Modified = value
        End Set
    End Property

    Public DocumentPath As String = ""
    Public ProgramVersion As String = ProgramVersion
    Public NTSpecificationVersion As Decimal = NTVersion

    Public Title As String = ""
    Public Language As String = ""
    Public Author As String = ""
    Public Website As String = ""
    Public CustomSymbols As String = ""
    Public Info As String = ""

    Public Pages As New List(Of NotebookPage)
    Public WordDictionary As New DictionaryFile

    Public Sub Save(ByVal FilePath As String)
        If File.Exists(FilePath) Then File.Delete(FilePath)
        NotebookFileAccess.Save(FilePath, Me)
        Modified = False
    End Sub

    Public Sub Open(ByVal FilePath As String, Optional ByVal Opening As Boolean = False)
        Dim NewNotebook As NotebookFile = NotebookFileAccess.Open(FilePath)
        Author = NewNotebook.Author
        ProgramVersion = NewNotebook.ProgramVersion
        NTSpecificationVersion = NewNotebook.NTSpecificationVersion
        Title = NewNotebook.Title
        Language = NewNotebook.Language
        Website = NewNotebook.Website
        Pages = NewNotebook.Pages
        DocumentPath = FilePath
        Info = NewNotebook.Info
        CustomSymbols = NewNotebook.CustomSymbols
        WordDictionary = NewNotebook.WordDictionary
    End Sub

    Public Sub New()
    End Sub
End Class

Module NotebookFileAccess
    Dim PagesFolder As String
    Sub Save(ByVal FilePath As String, ByVal Notebook As NotebookFile)
        MainForm.SaveTabs()

        If DictionaryForm.Loaded = False Then
            DictionaryForm.LoadDictionary()
        End If

        DictionaryForm.SaveDictionary()

        Dim guid As Guid = Guid.NewGuid
        Dim tmp As String = Application.LocalUserAppDataPath & "\zip-" & guid.ToString
        If Directory.Exists(tmp) Then
            Directory.Delete(tmp, True)
        End If

        Directory.CreateDirectory(tmp)

        PagesFolder = tmp & "\pages\"
        Directory.CreateDirectory(PagesFolder)

        Dim DataFile As New List(Of ZiaLine) From {
            New ZiaLine(LineType.Comment, "Settings"),
            New ZiaLine(LineType.KeyValue, "Title", Notebook.Title),
            New ZiaLine(LineType.KeyValue, "Language", Notebook.Language),
            New ZiaLine(LineType.KeyValue, "Author", Notebook.Author),
            New ZiaLine(LineType.KeyValue, "Website", Notebook.Website),
            New ZiaLine(LineType.KeyValue, "NTVersion", NTVersion),
            New ZiaLine(LineType.KeyValue, "LangPadVersion", LangPadVersion),
            New ZiaLine(LineType.Blank),
            New ZiaLine(LineType.Comment, "Pages")
        }

        For i = 0 To Notebook.Pages.Count - 1
            Dim Page = Notebook.Pages.Item(i)
            DataFile.Add(New ZiaLine(LineType.KeyValue, "Page" & i, ToCompatibleString(Page.Title)))

            Dim txtWriter As StreamWriter
            txtWriter = New StreamWriter(PagesFolder & i & ".rtf")
            txtWriter.Write(Page.RTF)
            txtWriter.Close()
        Next

        'Write to disk
        File.WriteAllText(tmp & "\data.txt", Write(DataFile))
        File.WriteAllText(tmp & "\info.txt", Notebook.Info)
        File.WriteAllText(tmp & "\custom_symbols.txt", Notebook.CustomSymbols)
        Notebook.WordDictionary.Save(tmp & "\dictionary.txt")
        Notebook.DocumentPath = FilePath

        ZipFile.CreateFromDirectory(tmp, FilePath, CompressionLevel.Optimal, False)
    End Sub

    Public Function Open(ByVal FilePath As String) As NotebookFile
        Dim NewNotebook As New NotebookFile
        Dim guid As Guid = Guid.NewGuid
        Dim tmp As String = Application.LocalUserAppDataPath & "\zip-" & guid.ToString

        If (Directory.Exists(tmp)) Then
            Directory.Delete(tmp, True)
        End If

        Directory.CreateDirectory(tmp)

        ZipFile.ExtractToDirectory(FilePath, tmp)
        PagesFolder = tmp & "\pages\"

        Dim LineList = Read(File.ReadAllText(tmp & "\data.txt"))
        NewNotebook.NTSpecificationVersion = Search(LineList, "NTVersion")
        Dim CompatMode As Boolean = (NewNotebook.NTSpecificationVersion < 2)

        If CompatMode Then ' Use the legacy method of storing pages
            Dim PageOrder = Search(LineList, "Page Order").Split({"|"}, StringSplitOptions.RemoveEmptyEntries)
            For Each PageTitle As String In PageOrder
                Dim Page As New NotebookPage With {
                    .Title = PageTitle,
                    .RTF = File.ReadAllText(PagesFolder & PageTitle & ".rtf")
                }
                NewNotebook.Pages.Add(Page)
            Next
        Else ' Use the new method of storing pages that allows for arbitrary names
            Dim PageCount As Integer = Directory.EnumerateFiles(PagesFolder).Count
            For i = 0 To PageCount - 1
                Dim Page As New NotebookPage
                Dim PageName As String = Search(LineList, "Page" & i)

                If PageName Is Nothing Then
                    PageName = i + 1
                End If

                Page.Title = PageName
                Page.RTF = File.ReadAllText(PagesFolder & i & ".rtf")
                NewNotebook.Pages.Add(Page)
            Next
        End If

        NewNotebook.Title = Search(LineList, "Title")
        NewNotebook.Language = Search(LineList, "Language")
        NewNotebook.Author = Search(LineList, "Author")
        NewNotebook.Website = Search(LineList, "Website")
        NewNotebook.ProgramVersion = Search(LineList, "LangPadVersion")
        NewNotebook.CustomSymbols = File.ReadAllText(tmp & "\custom_symbols.txt")
        NewNotebook.Info = File.ReadAllText(tmp & "\info.txt")
        NewNotebook.WordDictionary.Open(tmp & "\dictionary.txt")

        If NewNotebook.CustomSymbols <> "" Then
            CharEditor.charEdit.FilePanel.Controls.Clear()

            Dim FileChars As String() = NewNotebook.CustomSymbols.Split({Environment.NewLine}, StringSplitOptions.RemoveEmptyEntries)
            For Each FileChar As String In FileChars
                CharEditor.charEdit.InsertCharacterButton(FileChar, CharEditor.charEdit.FilePanel)
            Next
        End If

        Return NewNotebook
    End Function
End Module

File: TundraLib/Notebook/NotebookFile.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.IO.Compression;
using static TundraLib.ZiaFile;

namespace TundraLib.Notebook
{
    [Serializable()]
    public class NotebookPage
    {
        public string Title;
        public string RTF;
    }

    [Serializable()]
    public class NotebookFile
    {
        public bool Modified = false;
        public bool EmbedSymbols = false;
        public string DocumentPath;
        public string ProgramVersion = ProgramVersion;
        public decimal NTSpecificationVersion = NTVersion;

        public string Title;
        public string Language;
        public string Author;
        public string Website;
        public string CustomSymbols;
        public string Info;

        public List<NotebookPage> Pages = new List<NotebookPage>();
        public DictionaryFile WordDictionary = new DictionaryFile();

        public void Save(string FilePath)
        {
            if (File.Exists(FilePath))
                File.Delete(FilePath);
            NotebookFileAccess.Save(FilePath, this);
            Modified = false;
        }

        public void Open(string FilePath, bool Opening = false)
        {
            NotebookFile NewNotebook = NotebookFileAccess.Open(FilePath);
            Author = NewNotebook.Author;
            ProgramVersion = NewNotebook.ProgramVersion;
            NTSpecificationVersion = NewNotebook.NTSpecificationVersion;
            Title = NewNotebook.Title;
            Language = NewNotebook.Language;
            Website = NewNotebook.Website;
            Pages = NewNotebook.Pages;
            DocumentPath = FilePath;
            Info = NewNotebook.Info;
            CustomSymbols = NewNotebook.CustomSymbols;
            EmbedSymbols = NewNotebook.EmbedSymbols;
            WordDictionary = NewNotebook.WordDictionary;
        }

        public NotebookFile()
        {
        }
    }

    static class NotebookFileAccess
    {
        private static string PagesFolder;
        public static void Save(string FilePath, NotebookFile Notebook)
        {
            frmMain.SaveTabs();

            if (frmDictionary.Loaded == false)
                frmDictionary.LoadDictionary();

            frmDictionary.SaveDictionary();

            Guid guid = Guid.NewGuid();
            string tmp = Application.LocalUserAppDataPath + @"\zip-" + guid.ToString();
            if (Directory.Exists(tmp))
                Directory.Delete(tmp, true);

            Directory.CreateDirectory(tmp);

            PagesFolder = tmp + @"\pages\";
            Directory.CreateDirectory(PagesFolder);

            List<ZiaLine> DataFile = new List<ZiaLine>()
        {
            new ZiaLine(LineType.Comment, "Settings"),
            new ZiaLine(LineType.KeyValue, "Title", Notebook.Title),
            new ZiaLine(LineType.KeyValue, "Language", Notebook.Language),
            new ZiaLine(LineType.KeyValue, "Author", Notebook.Author),
            new ZiaLine(LineType.KeyValue, "Website", Notebook.Website),
            new ZiaLine(LineType.KeyValue, "NTVersion", NTVersion),
            new ZiaLine(LineType.KeyValue, "LangPadVersion", LangPadVersion),
            new ZiaLine(LineType.Blank),
            new ZiaLine(LineType.Comment, "Pages")
        };

            for (var i = 0; i <= Notebook.Pages.Count - 1; i++)
            {
                var Page = Notebook.Pages[i];
                DataFile.Add(new ZiaLine(LineType.KeyValue, "Page" + i, ToCompatibleString(Page.Title)));

                StreamWriter txtWriter;
                txtWriter = new StreamWriter(PagesFolder + i + ".rtf");
                txtWriter.Write(Page.RTF);
                txtWriter.Close();
            }

            if (Notebook.EmbedSymbols == true)
                Notebook.CustomSymbols = My.Settings.CustomSymbols;
            else
                Notebook.CustomSymbols = "";

            // Write to disk
            File.WriteAllText(tmp + @"\data.txt", Write(DataFile));
            File.WriteAllText(tmp + @"\info.txt", Notebook.Info);
            File.WriteAllText(tmp + @"\custom_symbols.txt", Notebook.CustomSymbols);
            Notebook.WordDictionary.Save(tmp + @"\dictionary.txt");
            Notebook.DocumentPath = FilePath;

            ZipFile.CreateFromDirectory(tmp, FilePath, CompressionLevel.Optimal, false);
        }

        public static NotebookFile Open(string FilePath)
        {
            NotebookFile NewNotebook = new NotebookFile();
            Guid guid = Guid.NewGuid();
            string tmp = Application.LocalUserAppDataPath + @"\zip-" + guid.ToString();

            if ((Directory.Exists(tmp)))
                Directory.Delete(tmp, true);

            Directory.CreateDirectory(tmp);

            ZipFile.ExtractToDirectory(FilePath, tmp);
            PagesFolder = tmp + @"\pages\";

            var LineList = Read(File.ReadAllText(tmp + @"\data.txt"));
            NewNotebook.NTSpecificationVersion = decimal.Parse(Search(LineList, "NTVersion"));
            bool CompatMode = (NewNotebook.NTSpecificationVersion < 2);

            if (CompatMode)
            {
                var PageOrder = Search(LineList, "Page Order").Split(
                {
                    "|"
            }, StringSplitOptions.RemoveEmptyEntries);
                foreach (string PageTitle in PageOrder)
                {
                    NotebookPage Page = new NotebookPage()
                    {
                        Title = PageTitle,
                        RTF = File.ReadAllText(PagesFolder + PageTitle + ".rtf")
                    };
                    NewNotebook.Pages.Add(Page);
                }
            }
            else
            {
                int PageCount = Directory.EnumerateFiles(PagesFolder).Count();
                for (var i = 0; i <= PageCount - 1; i++)
                {
                    NotebookPage Page = new NotebookPage();
                    string PageName = Search(LineList, "Page" + i);

                    if (PageName == null)
                        PageName = (i + 1).ToString();

                    Page.Title = PageName;
                    Page.RTF = File.ReadAllText(PagesFolder + i + ".rtf");
                    NewNotebook.Pages.Add(Page);
                }
            }



            NewNotebook.Title = Search(LineList, "Title");
            NewNotebook.Language = Search(LineList, "Language");
            NewNotebook.Author = Search(LineList, "Author");
            NewNotebook.Website = Search(LineList, "Website");
            NewNotebook.ProgramVersion = Search(LineList, "LangPadVersion");
            NewNotebook.CustomSymbols = File.ReadAllText(tmp + @"\custom_symbols.txt");
            NewNotebook.Info = File.ReadAllText(tmp + @"\info.txt");
            NewNotebook.WordDictionary.Open(tmp + @"\dictionary.txt");

            if (!(NewNotebook.CustomSymbols == ""))
                NewNotebook.EmbedSymbols = true;

            if (NewNotebook.CustomSymbols == "")
            {
            }
            else if (NewNotebook.CustomSymbols == My.Settings.CustomSymbols)
            {
            }
            else if (My.Settings.CustomSymbols.StartsWith(NewNotebook.CustomSymbols))
            {
            }
            else
            {
                dlgAppendReplace.ShowDialog();
                if (dlgAppendReplace.Result == dlgAppendReplace.AppendReplaceDialogResult.Append)
                    My.Settings.CustomSymbols = My.Settings.CustomSymbols + NewNotebook.CustomSymbols;
                else if (dlgAppendReplace.Result == dlgAppendReplace.AppendReplaceDialogResult.Replace)
                    My.Settings.CustomSymbols = NewNotebook.CustomSymbols;
                else
                {
                }

                frmMain.CustomLayoutPanel.Controls.Clear();
                frmDictionary.CustomLayoutPanel.Controls.Clear();

                string[] LineList2 = My.Settings.CustomSymbols.Split(
                {
                    Environment.NewLine
            }, StringSplitOptions.RemoveEmptyEntries);
                foreach (string IPA in LineList2)
                {
                    SymbolButton SymbolButton = new SymbolButton()
                    {
                        Text = IPA
                    };
                    SymbolButton.Click += frmMain.InsertIPA;
                    frmMain.CustomLayoutPanel.Controls.Add(SymbolButton);

                    SymbolButton SymbolButtonDictionary = new SymbolButton()
                    {
                        Text = IPA
                    };
                    SymbolButtonDictionary.Click += frmDictionary.InsertIPA;
                    frmDictionary.CustomLayoutPanel.Controls.Add(SymbolButtonDictionary);
                }
            }

            return NewNotebook;
        }
    }
}